### ISSUE TITLE
[Core Port Req.] PCSX2 - PlayStation 2

### ISSUE THREAD

**Issue by Spikestuff** [TASEmulators/BizHawk#3318]
### Upstream info
- [Website](https://pcsx2.net/)
- Target platforms: Win/Mac/Tux
- [Source repo](https://github.com/PCSX2/pcsx2)
- Language(s): C++
- License: GPL v3

### Merits
[//]: # "Briefly explain why this emulator is worth including in BizHawk. If it emulates the same system as an existing core, compare them."
PlayStation 2 is probably the one with the largest backlog, and the most potential for people to make runs and submit them.
It's insanely vary and I know that RTA runners of certain titles want to get in TASing a PS2 title, but haven't been able to due to how unstable upsream is, the [TAS fork](https://github.com/xTVaser/pcsx2-rr/releases/tag/rr-1.0.1) is stable by comparison.

### Technical details
Now for the annoying part.
Firstly DobieStation (experimental) has to go. It's been more than a year since the project was touched on by PSI and is basically abandoned, thankfully the code was compiled with those that made PCSX2 and they transferred what they learnt into PCSX2...

In the settings of PCSX2 (Config > General Settings) the EmotionEngine/IOP and VU0 and VU1 should be using their Recompiler modes. Using the Interpreter mode, tanks performance and is something that people would avoid using the product.
Their Advanced Options for Round and Clamping should be Chop/Zero and Full (Extra + Preserve Sign in VU) respectfully.
This will allow the game to actually be represented correctly. And this **should not** be modifiable.

In terms of Graphics, I'm glad I learnt this with the [Naruto TAS](https://youtu.be/ZJ5oaQwJV2M) which did use the aforementioned TAS fork and not upstream cause what it taught me is that **any renderer can be used**, and it will sync across the board hopefully more in practice can confirm that than in theory.
**Ideally** the Software Renderer is the one Hawk should prioritize default on but allowing Vulkan/DirectX/GL should be an option as well, to at least get performance on any game that does tax the Emulator on Software mode, and with the single renderer (it's actually very optimized in Software).

The Graphics settings in Software Mode should all be enabled by default without being able to turn them off cause well to keep simple they're actually modes that get the games that need them to render correctly from Grand Theft Auto: San Andreas to something as normal as Quake III: Revolution (You need that Mipmapping).

Lastly in an instance of having preferred choice of renderer.
Need for Speed: Carbon is actually missing the tint it has on Software Mode, but it's present in the Hardware renderers.
Cause on the opposite, fog is absolutely scuffed in Hardware mode for Crash: Wrath of Cortex where Software would be preferable, there's also texture issues in Hardware mode for Ridge Racer V for instance.
The outliers in getting it dumped and encoded correctly is another reason to have the choice in renderers.

As for the Deinterlacing. The options provided which are; None, Weave (bff) and Bob (bff) would be preferable.

**Comment by CasualPokePlayer** [comment]
If this is to be waterboxed, Vulkan/DirectX/GL will not be options (native GPU stuff is not possible in the box), it will have to be software rendering. Single threading would also need to be forced probably (this is configurable so I assume there isn't much work there).

Recompilers might also be issues too if it is to be waterboxed. In theory it should work, in practice it can be messy depending on the implementation. A quick look seems to say at least they don't use fastmem, which probably means it's possible. If it turns out to not be possible, and interpreters have to be used, then waterbox is probably not viable.

Which even then, if this is a native port, how much more stable would BizHawk actually be compared to upstream? We'd be relying on their states, which seem to not be reliable in the first place. At most you can get BizHawk's tools, but you'd still be in desync hell.

**Comment by nattthebear** [comment]
I don't think waterbox is viable here.  Most JITs are too strenuous for it, and (real) multithreading is a no-go as well.

**Comment by InfamousKnight** [comment]
What is so unstable about upstream I'm curios? Wouldn't it have less desyncs than that aforementioned tas fork? Because while I was following the development of pcsx2, they updated save state code to "hopefully fix desync problems". And sometimes desyncs are frontend related. Plus they are moving away from hacky emulation fixes and have removed several hacks in the past if you followed their progress.

Whats really great is they have compressed save states now which can make rewind more reasonable. They dont want to implement the feature though because of performance problems.

**Comment by InfamousKnight** [comment]
Yeah, well that's game specific. I've heard games like ratchet and clank are pretty stable(desyncs happen between transition to another planet) so it's good for individual levels. I've also heard that for kingdom hearts ii, certain actions had to be avoided to prevent desyncs.

Its very complex.

**Comment by CasualPokePlayer** [comment]
That further supports it's state related.

**Comment by InfamousKnight** [comment]
One more thing to add, those reports listed above are rather old. The kingdom hearts tas I was talking about where certain actions were avoided was made in 2020. Back in October 2021, a massive save state improvement was addressed targeting desync issues with missing variables.

Unfortunately for upstream, because of their move to qt, I don't think I can test anything on their end for any up to date info until qt is done on the recording tools side.

**Comment by buruto** [comment]
wxWidgets has been completely removed and the next stable release of PCSX2 is arriving soon and it will be completely Qt.

Since PCSX2 is open source, isn't it possible to create a pcsx2.dll with exported functions that can emulate the PS2 hardware and BizHawk will load and use it to play PS2 games on windows?

You can also use the PCSX2 code that saves and load states, make an export function "SaveState" and another export function "LoadState" and BizHawk will simply invoke them?

**Comment by CasualPokePlayer** [comment]
It's a lot easier said than done to just "build a dll" especially when an emulator is not at all intended to be used in such a way. And whether the standalone emu uses wx or Qt is of little concern of us, as such a dll would be stripped of the frontend.

Also, the way you say this implies we would consider having a windows only core for PS2. This is unacceptable, Linux should be just as supported, and it is even more unacceptable given pcsx2 works fine on Linux.

Also if you were not aware, simply having savestate support (which in any case is mandatory for releasing a BizHawk core) automatically means rewind support, as the rewinder is implemented purely in the frontend, the core is only responsible for actual savestating.

**Comment by buruto** [comment]
> It's a lot easier said than done to just "build a dll" especially when an emulator is not at all intended to be used in such a way. And whether the standalone emu uses wx or Qt is of little concern to us, as such a dll would be stripped of the frontend (same goes for your remarks about a debugger, as that is largely part of the frontend).
> 
> Also, the way you say this implies we would consider having a windows only core for PS2. This is unacceptable, Linux should be just as supported, and it is even more unacceptable given pcsx2 works fine on Linux.
> 
> Also if you were not aware, simply having savestate support (which in any case is mandatory for releasing a BizHawk core) automatically means rewind support, as the rewinder is implemented purely in the frontend, the core is only responsible for actual savestating.

So playing PS2 with BizHawk is impossible?

**Comment by CasualPokePlayer** [comment]
It is impossible *at the moment*, it may be possible with time. Please do not expect instant results here, we're ultimately just a small group of volunteers, and we already have other things to do.

**Comment by buruto** [comment]
> It is impossible _at the moment_, it may be possible with time. Please do not expect instant results here, we're ultimately just a small group of volunteers, and we already have other things to do.

What has to be implemented in order to make this possible?

And I am used to the fact that I have to wait years to get something done.

**Comment by CasualPokePlayer** [comment]
Wiring up everything and any frontend and core separating is not a quick a trivial task.

**Comment by buruto** [comment]
> Wiring up everything and any frontend and core separating is not a quick a trivial task.

I did not say anything about the complexity and difficulty of doing this. I just wanted to know if this is possible to do it this way.

**Comment by CasualPokePlayer** [comment]
Yes, in a very generic way that is one way to describe the process. That doesn't simplify things at all, as you are just ultimately describing a very generic version of the porting process that ignores nuances.

**Comment by buruto** [comment]
> Yes, in a very generic way that is one way to describe the process. That doesn't simplify things at all, as you are just ultimately describing a very generic version of the porting process that ignores nuances.

What nuances exactly?

**Comment by KinadaDragon** [comment]
Any updates on this core being added? Would love to do a THPS series shuffle playthrough, but anywhere I look I cannot find a viable option on said idea.
